ql/src/java/org/apache/hadoop/hive/ql/ddl/function/AbstractFunctionAnalyzer.java

```diff
@@ -66,11 +70,24 @@ protected void addEntities(String functionName, String className, boolean isTemp
     }
     if (database != null) {
       outputs.add(new WriteEntity(database, WriteEntity.WriteType.DDL_NO_LOCK));
+      // Add the permanent function as a WriteEntity
+      Function function;
+      if (isCreate) {
+        function = new Function(functionName, database.getName(), className,
+                SessionState.getUserFromAuthenticator(), PrincipalType.USER,
+                (int) (System.currentTimeMillis() / 1000), FunctionType.JAVA, resources);
+      } else {
+        try {
+          function = db.getFunction(database.getName(), functionName);
```

Comment:
Why we fetch the function here, can we just create the function as the temporary function does?

Reply:
We need to fetch the function for desc or drop function queries, we need to know whether the current user has required permissions by looking at the function object.

Reply:
Yeah that's correct